### DISCUSSION TITLE
Possibility to set Kernel path manually?

### DISCUSSION THREAD

**Discussion by Tma2333** [microsoft/vscode-jupyter#9693] [Questions and Answers] [answered]
I am developing on shared server with network attached drive and large amount of conda environment, the `Discovering Python Interpreters` process could take hours to complete. My only saving grace was the last used kernel will be saved? (I found there is some information being saved in `temp`). Is it possible to cache more than one Kernel? And, what process would trigger caching?  

Is there any way we can manually give the path to the interpreter, rather than wish Discovering would be done quickly or the correct interpreter shows up eventually.

**Comment by IanMatthewHuff** [comment, marked as answer]
@Tma2333 Just to check, the situation that you are seeing is that there is one conda environment being cached that you can select and use before the `Discovering Python Interpreters` task is finished? While the process is slow (I believe the Python extension team is working on this) after an initial load the local kernel connections should all be cached and subsequent returns should be quite speedy. Could you possibly save off the content of the Jupyter tab of the output window? If you save that in a file and attach it to a comment here (easier to attach a file versus pasting the whole thing) I could take a peek and see if something is up with the caching.

**Reply by Tma2333** [reply]
Since everything seem to be cache, and `Discovering Python Interpreters` process is finished, it is hard to reproduce this behavior. 

Yes, currently, I am only use one of the conda environment. And after startup (ssh remotely into the server), there is one conda environment I can used before `Discovering Python Interpreters`. Is this caching behavior tie to specific notebook file or the extension? i.e. If I open `A.ipynb` and use `conda:A` and open `B.ipynb` and use `conda:B` will both be cached? 

Also, another thing I saw is that, I have load two environment using the extension i.e. open `A.ipynb` and use `conda:A` and open `B.ipynb` and use `conda:B`. I am not sure if this is where kernel being cache, in `~/.vscode-server/extensions/ms-toolsai.jupyter-2022.3.1000901801/temp/jupyter/kernels/python397jvsc74a57bd0fe50492733d7449d3e23ddde50178f9615d673c5e7dc001510615bbd620e525d/kernel.json` I only see one environment. Or in $~/.vscode-server/extensions/ms-toolsai.jupyter-2022.3.1000901801/temp/jupyter/kernels/`, I only saw one directory . 

Sorry, not sure which file you mean to ask for, do you need the `.ipynb` notebook file? Thank you for the help!

**Reply by IanMatthewHuff** [reply]
So the file I was asking for was the contents of this window here (specifically the jupyter tab):
<img width="1405" alt="image" src="https://user-images.githubusercontent.com/812783/163623900-074dcac5-7f2a-42e5-8284-8b89ab13cb8d.png">
It's actually not a file. But the contents here can be really long, so it's easier to save and attach it as a text file versus pasting the contents in. We have some logging in here that could indicate to me if the caching is failing.

**Reply by IanMatthewHuff** [reply]
The caching should cache all of the conda environments that are on the system. I'd expect that the full set of them would be available the next time.

**Reply by Tma2333** [reply]
[jupyter_output.txt](https://github.com/microsoft/vscode-jupyter/files/8497933/jupyter_output.txt)

Here it is!

**Reply by IanMatthewHuff** [reply]
Thanks. I'll look over this some today. At first blush I think that I misspoke regarding one thing. While I do expect the full set of environments to be cached we do also cache the specific preferred kernel individually. To some degree we are dealing with a bit of known issue here. The first time it's run it's possible that discovering python environments could take a very long time. Though I do believe that there is a specific issue that is getting addressed here. @rchiodo were you looking into a python discovery slowdown on the build machines?

**Reply by Tma2333** [reply]
Thank you! I opened a new server instance and even if `Discovering Python Interpreters` is running, I still have access to full set of environment. It seems a bit hard to reproduce the issues I was seeing since everything is caching properly. 

The issues I had before was there only 1 environment (the last run one) showing up rather than full set of them. If it came up again, I will make sure comment it further here and post corresponding log.
